### ISSUE TITLE
Improve the algorithm used to connect endpoints

### ISSUE THREAD

**Issue by ShravanDoda** [HTTP-APIs/python-hydra-agent#83]

**Comment by Mec-iS** [comment]
we are working with the perspective of storing hundreds of millions of nodes, that is why we use Redis. Graph traversal has to be done using built-in Redis data structures and [redis-graph querying system](https://oss.redislabs.com/redisgraph/)

**Comment by ShravanDoda** [comment]
@Mec-iS the issue is not with how edges are constructed between 2 nodes. It's with how the nodes to be connected are identified in the agent. Sorry for improper language.

**Comment by Mec-iS** [comment]
in any case we cannot keep a dictionary in memory with nodes data. Redis is already an in-memory datastore, anything about the graph should be queried from the graph.

**Comment by ShravanDoda** [comment]
The way program works right now is that it uses a function to create nodes and then connect them. How about we use a dictionary as a local variable in the function? Using 4 nested loops is going to be very inefficient especially as the number of nodes increase.

**Comment by Mec-iS** [comment]
@sandeep1 ^

**Comment by Mec-iS** [comment]
@sandeepsajan0 ^

**Comment by Mec-iS** [comment]
> Using 4 nested loops is going to be very inefficient especially as the number of nodes increase.

yeah this true. but any query or operation on the graph has to be done in Redis, we cannot allow to have the graph/nodes replicated by Python memory at any time. So I would try to refactor that code by leveraging [Redis commands](https://redis.io/commands). The first version of the agent will be strictly coupled to Redis, working on it requires an understanding about what the datastore can do.
